File: 动态规划系列/coin.py
#!/usr/bin/python
# -*- coding: UTF-8 -*-

def coinChange(coins, amount):

    def dp(n):
        # base case
        if n == 0: return 0
        if n < 0: return -1
        # 求最小值，所以初始化为正无穷
        res = float('INF')
        for coin in coins:
            subproblem = dp(n - coin)
            # 子问题无解，跳过
            if subproblem == -1: continue
            res = min(res, 1 + subproblem)

        return res if res != float('INF') else -1
    
    return dp(amount)

def coinChange2(coins, amount):
    # 备忘录
    memo = dict()
    
    def dp(n):
        # 查备忘录，避免重复计算
        if n in memo: return memo[n]

        if n == 0: return 0
        if n < 0: return -1
        res = float('INF')
        for coin in coins:
            subproblem = dp(n - coin)
            if subproblem == -1: continue
            res = min(res, 1 + subproblem)
        
        # 记入备忘录
        memo[n] = res if res != float('INF') else -1
        return memo[n]
    
    return dp(amount)

ret = coinChange2([1, 2, 5], 11)
print(ret)

File: 动态规划系列/coin.cpp
#include <vector>
#include <cstdio>

using namespace std;

int coinChange(vector<int>& coins, int amount) {
    // 数组大小为 amount + 1，初始值也为 amount + 1
    vector<int> dp(amount + 1, amount + 1);
    // base case
    dp[0] = 0;
    for (int i = 0; i < dp.size(); i++) {
        // 内层 for 在求所有子问题 + 1 的最小值
        for (int coin : coins) {
            // 子问题无解，跳过
            if (i - coin < 0) continue;
            dp[i] = min(dp[i], 1 + dp[i - coin]);
        }
    }
    return (dp[amount] == amount + 1) ? -1 : dp[amount];
}


int main(int argc, char const *argv[])
{
    vector<int> coins(3, 0);
    coins[0] = 2;
    coins[1] = 4;
    coins[2] = 5;

    int ret = coinChange(coins, 11);
    printf("coinChange(11) = %d\n", ret);
    return 0;
}
